File: pipela_qt/resolution_chrome.py
"""클라·템플릿·DPI 한 줄(리치 텍스트) — 제어창·게임 상단 스트립 공통."""

from __future__ import annotations

from dataclasses import dataclass

from PyQt6.QtGui import QFont, QTextDocument
from PyQt6.QtWidgets import QLabel

from pipela_qt import theme as T
from pipela_qt.ui_adaptive import root_font_pt, scale_px_h, scale_px_v


@dataclass(frozen=True)
class ResolutionHtmlPalette:
    muted: str
    val: str
    warn: str
    sep: str
    note: str


CONTROL_RESOLUTION_PALETTE = ResolutionHtmlPalette(
    muted=T.FG_DIM,
    val=T.ACCENT,
    warn=T.STATUS_WARN,
    sep=T.BORDER_HAIR,
    note=T.FG_MUTED,
)

STRIP_RESOLUTION_PALETTE = ResolutionHtmlPalette(
    muted=T.STRIP_FG_MUTED,
    val=T.STRIP_ACCENT,
    warn=T.STATUS_WARN,
    sep=T.STRIP_BORDER,
    note=T.STRIP_FG_MUTED,
)


def _windows_dpi_parts(m, hwnd) -> tuple[int, int]:
    """
    창이 올라간 **모니터**의 배율(shcore.GetDpiForMonitor effective).
    hwnd가 없으면 Pipela 제어창 HWND(`pipela_qt_control_win_hwnd`)로 그 모니터를 쓴다.
    반환: (퍼센트 정수, DPI 정수).
    """
    try:
        ref = int(hwnd) if hwnd else getattr(m, "pipela_qt_control_win_hwnd", None)
        if ref:
            d = int(m.get_dpi_for_monitor_containing_window(ref))
        else:
            d = int(m.get_native_window_dpi(None))
    except Exception:
        d = 96
    try:
        pct = int(round(d / 96.0 * 100.0))
    except Exception:
        pct = 100
    return pct, d


def _res_metrics_sep(p: ResolutionHtmlPalette) -> str:
    return f'<span style="color:{p.sep};">·</span>'


def _res_lbl_muted(t: str, p: ResolutionHtmlPalette) -> str:
    return f'<span style="color:{p.muted};">{t}</span>'


def _res_lbl_val(t: str, p: ResolutionHtmlPalette) -> str:
    return f'<span style="color:{p.val}; font-weight:600;">{t}</span>'


def _res_lbl_warn(t: str, p: ResolutionHtmlPalette) -> str:
    return f'<span style="color:{p.warn}; font-weight:600;">{t}</span>'


def _res_tpl_note_html(p: ResolutionHtmlPalette) -> str:
    return (
        f'<span style="color:{p.note}; font-size:92%; font-weight:400;">'
        f"&nbsp;(1440p)"
        f"</span>"
    )


def _dpi_fragment_html(m, hwnd, p: ResolutionHtmlPalette) -> str:
    """% 와 유효 DPI 정수는 한 덩어리로 붙여 표시(구분 `·` 없음, 좁은 공백만)."""
    pct, d = _windows_dpi_parts(m, hwnd)
    _tight = "&#8239;"
    return (
        f"{_res_lbl_muted('DPI', p)}&nbsp;{_res_lbl_val(f'{pct}%', p)}{_tight}"
        f"{_res_lbl_val(str(d), p)}"
    )


def resolution_main_and_dpi_html(
    m,
    p: ResolutionHtmlPalette,
) -> tuple[str, str]:
    """(클라·템플릿 등 본문 HTML, DPI HTML). 오류 시 (한 덩어리, '')."""
    sep = _res_metrics_sep(p)
    try:
        target_hwnd = m.refresh_target_hwnd_if_needed()
        if target_hwnd:
            if m.is_window_minimized(target_hwnd):
                return (
                    f"{_res_lbl_warn('게임 최소화', p)}{sep}"
                    f"{_res_lbl_muted('캡처·매칭 대기', p)}",
                    _dpi_fragment_html(m, target_hwnd, p),
                )
            sz = m.get_window_size(target_hwnd)
            if sz:
                gw, gh = sz
                r = float(gh) / float(m.BASE_HEIGHT)
                note = _res_tpl_note_html(p)
                return (
                    f"{_res_lbl_muted('클라', p)}&nbsp;{_res_lbl_val(f'{gw}×{gh}', p)}{sep}"
                    f"{_res_lbl_muted('템플릿', p)}&nbsp;{_res_lbl_val(f'{r:.3f}', p)}{note}",
                    _dpi_fragment_html(m, target_hwnd, p),
                )
        _luh = m.refresh_smart_updater_hwnd_if_needed()
        if _luh and not m.is_window_minimized(_luh):
            sz_l = m.get_window_size(_luh)
            if sz_l:
                lw, lh = sz_l
                r = float(lh) / float(m.BASE_HEIGHT)
                note = _res_tpl_note_html(p)
                return (
                    f"{_res_lbl_muted('클라', p)}&nbsp;{_res_lbl_val(f'{lw}×{lh}', p)}{sep}"
                    f"{_res_lbl_muted('템플릿', p)}&nbsp;{_res_lbl_val(f'{r:.3f}', p)}{note}",
                    _dpi_fragment_html(m, _luh, p),
                )
            return (
                f"{_res_lbl_muted('연결', p)}&nbsp;{_res_lbl_val('런처만', p)}",
                _dpi_fragment_html(m, _luh, p),
            )
        return (
            f"{_res_lbl_muted('이터널시티 창 없음', p)}",
            _dpi_fragment_html(m, None, p),
        )
    except Exception:
        return (
            f'<span style="color:{p.muted};">해상도 정보 오류</span>',
            "",
        )


def resolution_block_html(m, p: ResolutionHtmlPalette) -> str:
    """클라·템플릿·DPI 지표를 한 줄(리치 텍스트)로 표시."""
    sep = _res_metrics_sep(p)
    main, dpi = resolution_main_and_dpi_html(m, p)
    if not dpi:
        return main
    return f"{main}{sep}{dpi}"


def resolution_block_content_key(m) -> tuple:
    """`resolution_block_html` 결과가 바뀔 때만 달라지는 값 — HTML 문자열 없이 저비용 비교용."""
    try:
        target_hwnd = m.refresh_target_hwnd_if_needed()
        if target_hwnd:
            if m.is_window_minimized(target_hwnd):
                return ("gmin", int(target_hwnd))
            sz = m.get_window_size(target_hwnd)
            if sz:
                gw, gh = int(sz[0]), int(sz[1])
                r = round(float(gh) / float(m.BASE_HEIGHT), 4)
                try:
                    d = int(m.get_dpi_for_monitor_containing_window(int(target_hwnd)))
                except Exception:
                    d = 96
                return ("game", int(target_hwnd), gw, gh, r, d)
        _luh = m.refresh_smart_updater_hwnd_if_needed()
        if _luh and not m.is_window_minimized(_luh):
            sz_l = m.get_window_size(_luh)
            if sz_l:
                lw, lh = int(sz_l[0]), int(sz_l[1])
                r = round(float(lh) / float(m.BASE_HEIGHT), 4)
                try:
                    d = int(m.get_dpi_for_monitor_containing_window(int(_luh)))
                except Exception:
                    d = 96
                return ("l_sz", int(_luh), lw, lh, r, d)
            try:
                d = int(m.get_dpi_for_monitor_containing_window(int(_luh)))
            except Exception:
                d = 96
            return ("l_only", int(_luh), d)
        ref = getattr(m, "pipela_qt_control_win_hwnd", None)
        try:
            if ref:
                d = int(m.get_dpi_for_monitor_containing_window(int(ref)))
            else:
                d = int(m.get_native_window_dpi(None))
        except Exception:
            d = 96
        return ("none", d)
    except Exception:
        return ("err",)


_RES_FIT_LETTER_SPACING_EM: tuple[float, ...] = (
    0.03,
    0.02,
    0.01,
    0.0,
    -0.015,
    -0.03,
    -0.045,
    -0.06,
    -0.075,
    -0.09,
)
_RES_FIT_BASE_DESIGN_PT = 10.5
_RES_FIT_MIN_DESIGN_PT = 3.85
_RES_FIT_ABSOLUTE_MIN_DESIGN_PT = 1.55
_RES_FIT_PT_STEP = 0.35
_RES_FIT_PT_STEP_FINE = 0.12


def _res_line_wrap(inner_html: str, letter_spacing_em: float) -> str:
    l = float(letter_spacing_em)
    return (
        '<div style="white-space:nowrap; '
        f'letter-spacing:{l:.3f}em;">{inner_html}</div>'
    )


def _measure_res_rich_width(font: QFont, html: str) -> float:
    doc = QTextDocument()
    doc.setDefaultFont(font)
    doc.setHtml(html)
    doc.setTextWidth(-1)
    return float(doc.idealWidth())


def apply_resolution_rich_label_fit(
    lbl: QLabel,
    m,
    avail_css_px: float,
    *,
    palette: ResolutionHtmlPalette,
    design_scale: float = 1.0,
    block_html: str | None = None,
) -> None:
    """가용 폭 안에 **항상 한 줄** — 필요 시 pt·자간 축소.

    ``block_html`` 이 있으면 ``resolution_main_and_dpi_html`` 재호출 없이 그대로 피팅(호출부에서
    ``resolution_block_html`` 과 짝지을 것).
    """
    if block_html is not None:
        single_inner = block_html
        if not single_inner:
            return
    else:
        main, dpi = resolution_main_and_dpi_html(m, palette)
        if not main and not dpi:
            return
        sep = _res_metrics_sep(palette)
        single_inner = f"{main}{sep}{dpi}" if dpi else main
    avail = max(40.0, float(avail_css_px))
    avail_fit = max(40.0, avail - float(scale_px_v(4)))
    avail_i = max(40, int(round(avail)))
    _rf = round(float(root_font_pt()), 3)
    _fit_k = (single_inner, avail_i, round(float(design_scale), 4), _rf)
    if getattr(lbl, "_pipela_res_fit_cache_k", None) == _fit_k:
        return
    # 가용 폭은 **부모(스트립의 _res_fill 등)에서 넘긴 값** — lbl.width() 로 잡지 말 것(자기축소 루프).
    try:
        lbl.setMaximumWidth(avail_i)
    except Exception:
        pass
    fit_limit = max(24.0, avail_fit * 0.99)
    scale = (float(root_font_pt()) / 11.0) * float(design_scale)
    base_pt = _RES_FIT_BASE_DESIGN_PT * scale
    coarse_floor = _RES_FIT_MIN_DESIGN_PT * scale
    abs_floor = _RES_FIT_ABSOLUTE_MIN_DESIGN_PT * scale
    chosen_pt = float(abs_floor)
    chosen_lsem = float(_RES_FIT_LETTER_SPACING_EM[-1])
    found = False
    pt = float(base_pt)
    # 넓으면 기본 pt·첫 자간에서 이미 맞는 경우가 많음 — QTextDocument 측정 횟수 대폭 감소
    try:
        _f_quick = QFont(lbl.font())
        _f_quick.setPointSizeF(pt)
        _w_quick = _res_line_wrap(single_inner, float(_RES_FIT_LETTER_SPACING_EM[0]))
        if _measure_res_rich_width(_f_quick, _w_quick) <= fit_limit:
            chosen_pt, chosen_lsem = pt, float(_RES_FIT_LETTER_SPACING_EM[0])
            found = True
    except Exception:
        pass
    while not found and pt >= coarse_floor - 1e-9:
        for lsem in _RES_FIT_LETTER_SPACING_EM:
            fnt = QFont(lbl.font())
            fnt.setPointSizeF(pt)
            wrapped = _res_line_wrap(single_inner, lsem)
            if _measure_res_rich_width(fnt, wrapped) <= fit_limit:
                chosen_pt, chosen_lsem = pt, float(lsem)
                found = True
                break
        if found:
            break
        pt -= _RES_FIT_PT_STEP
    if not found:
        pt = min(float(base_pt), coarse_floor)
        while pt >= abs_floor - 1e-9:
            for lsem in _RES_FIT_LETTER_SPACING_EM:
                fnt = QFont(lbl.font())
                fnt.setPointSizeF(pt)
                wrapped = _res_line_wrap(single_inner, lsem)
                if _measure_res_rich_width(fnt, wrapped) <= fit_limit:
                    chosen_pt, chosen_lsem = pt, float(lsem)
                    found = True
                    break
            if found:
                break
            pt -= _RES_FIT_PT_STEP_FINE
    if not found:
        chosen_pt = float(abs_floor)
        chosen_lsem = float(_RES_FIT_LETTER_SPACING_EM[-1])
    fnt = QFont(lbl.font())
    fnt.setPointSizeF(chosen_pt)
    lbl.setFont(fnt)
    lbl.setText(_res_line_wrap(single_inner, chosen_lsem))
    try:
        mh = lbl.fontMetrics().height()
        lbl.setMinimumHeight(max(scale_px_v(8), mh))
    except Exception:
        pass
    lbl._pipela_res_fit_cache_k = _fit_k


def apply_resolution_rich_label_fixed(
    lbl: QLabel,
    *,
    block_html: str,
    design_scale: float = 1.0,
) -> None:
    """스트립 등 — 적응형 축소 없이 고정 pt·자간 0으로 한 줄 리치 텍스트."""
    single_inner = (block_html or "").strip()
    if not single_inner:
        return
    _rf = round(float(root_font_pt()), 3)
    _fit_k = ("fixed", single_inner, round(float(design_scale), 4), _rf)
    if getattr(lbl, "_pipela_res_fit_cache_k", None) == _fit_k:
        return
    try:
        lbl.setMaximumWidth(16777215)
    except Exception:
        pass
    scale = (float(root_font_pt()) / 11.0) * float(design_scale)
    chosen_pt = float(_RES_FIT_BASE_DESIGN_PT * scale)
    fnt = QFont(lbl.font())
    fnt.setPointSizeF(chosen_pt)
    lbl.setFont(fnt)
    lbl.setText(_res_line_wrap(single_inner, 0.0))
    try:
        mh = lbl.fontMetrics().height()
        lbl.setMinimumHeight(max(scale_px_v(8), mh))
    except Exception:
        pass
    lbl._pipela_res_fit_cache_k = _fit_k

File: cpp/src/app/theme/resolution_chrome.cpp
#include "theme/resolution_chrome.hpp"

#include <cmath>

#include <QApplication>
#include <QFont>
#include <QLabel>
#include <QTextDocument>

#include "dock/dock_ui_phase.hpp"
#include "pipela/core/vision/roi.hpp"
#include "pipela/core/win32/game_windows.hpp"
#include "theme/dpi_helpers.hpp"
#include "theme/ui_adaptive.hpp"

namespace pipela::ui::theme {

namespace {

QString muted(const QString& text) {
    return QString::fromUtf8("<span style=\"color:#7a8a82;\">%1</span>").arg(text);
}

QString val(const QString& text) {
    return QString::fromUtf8("<span style=\"color:#3dd4c9; font-weight:600;\">%1</span>")
        .arg(text);
}

QString warn(const QString& text) {
    return QString::fromUtf8("<span style=\"color:#e8b060; font-weight:600;\">%1</span>")
        .arg(text);
}

QString sep() { return QString::fromUtf8("<span style=\"color:#2a3438;\">·</span>"); }

QString dpiFragment(std::intptr_t hwnd) {
    const double scale = win32DpiScaleForHwnd(hwnd);
    const int pct = static_cast<int>(std::lround(scale * 100.0));
    const int dpi = static_cast<int>(std::lround(scale * 96.0));
    return QString::fromUtf8("%1&nbsp;%2&#8239;%3")
        .arg(muted(QString::fromUtf8("DPI")))
        .arg(val(QString::fromUtf8("%1%").arg(pct)))
        .arg(val(QString::number(dpi)));
}

QString clientTemplateLine(int client_w, int client_h) {
    const double ratio =
        static_cast<double>(std::max(1, client_h)) / static_cast<double>(pipela::core::vision::kBaseHeight);
    return QString::fromUtf8("%1&nbsp;%2%3%4&nbsp;%5%6")
        .arg(muted(QString::fromUtf8("클라")))
        .arg(val(QString::fromUtf8("%1×%2").arg(client_w).arg(client_h)))
        .arg(sep())
        .arg(muted(QString::fromUtf8("템플릿")))
        .arg(val(QString::number(ratio, 'f', 3)))
        .arg(muted(QString::fromUtf8("&nbsp;(1440p)")));
}

std::pair<int, int> clientSizeLogical(std::intptr_t hwnd) {
    const auto cr = pipela::core::win32::getClientRectScreen(hwnd);
    return {std::get<2>(cr) - std::get<0>(cr), std::get<3>(cr) - std::get<1>(cr)};
}

QString resLineWrap(const QString& inner_html, double letter_spacing_em) {
    return QString::fromUtf8(
               "<div style=\"white-space:nowrap; letter-spacing:%1em;\">%2</div>")
        .arg(letter_spacing_em, 0, 'f', 3)
        .arg(inner_html);
}

double measureResRichWidth(const QFont& font, const QString& html) {
    QTextDocument doc;
    doc.setDefaultFont(font);
    doc.setHtml(html);
    doc.setTextWidth(-1);
    return static_cast<double>(doc.idealWidth());
}

constexpr double kResFitLetterSpacingEm[] = {
    0.03, 0.02, 0.01, 0.0, -0.015, -0.03, -0.045, -0.06, -0.075, -0.09,
};
constexpr double kResFitBaseDesignPt = 10.5;
constexpr double kResFitMinDesignPt = 3.85;
constexpr double kResFitAbsoluteMinDesignPt = 1.55;
constexpr double kResFitPtStep = 0.35;
constexpr double kResFitPtStepFine = 0.12;

}  // namespace

QString stripResolutionBlockHtml(std::intptr_t anchor_hwnd,
                                 std::intptr_t game_hwnd,
                                 std::intptr_t launcher_hwnd,
                                 pipela::ui::dock::UiDockPhase phase) {
    if (phase == pipela::ui::dock::UiDockPhase::Client && game_hwnd &&
        pipela::core::win32::isWindow(game_hwnd)) {
        if (pipela::core::win32::isWindowMinimized(game_hwnd)) {
            return warn(QString::fromUtf8("게임 최소화"))
                .append(sep())
                .append(muted(QString::fromUtf8("캡처·매칭 대기")))
                .append(sep())
                .append(dpiFragment(game_hwnd));
        }
        const auto [gw, gh] = clientSizeLogical(game_hwnd);
        if (gw > 0 && gh > 0) {
            return clientTemplateLine(gw, gh).append(sep()).append(dpiFragment(game_hwnd));
        }
    }
    if (launcher_hwnd && pipela::core::win32::isWindow(launcher_hwnd) &&
        !pipela::core::win32::isWindowMinimized(launcher_hwnd)) {
        const auto [lw, lh] = clientSizeLogical(launcher_hwnd);
        if (lw > 0 && lh > 0) {
            return clientTemplateLine(lw, lh).append(sep()).append(dpiFragment(launcher_hwnd));
        }
        return muted(QString::fromUtf8("연결"))
            .append(QString::fromUtf8("&nbsp;"))
            .append(val(QString::fromUtf8("런처만")))
            .append(sep())
            .append(dpiFragment(launcher_hwnd));
    }
    if (anchor_hwnd && pipela::core::win32::isWindow(anchor_hwnd)) {
        return muted(QString::fromUtf8("이터널시티 창 없음")).append(sep()).append(
            dpiFragment(anchor_hwnd));
    }
    return muted(QString::fromUtf8("이터널시티 창 없음"));
}

QString controlResolutionBlockHtml(std::intptr_t game_hwnd,
                                   std::intptr_t launcher_hwnd,
                                   pipela::ui::dock::UiDockPhase phase) {
    const std::intptr_t anchor =
        (phase == pipela::ui::dock::UiDockPhase::Launcher && launcher_hwnd) ? launcher_hwnd
                                                                              : game_hwnd;
    return stripResolutionBlockHtml(anchor, game_hwnd, launcher_hwnd, phase);
}

QString resolutionChromeContentKey(std::intptr_t anchor_hwnd,
                                   std::intptr_t game_hwnd,
                                   std::intptr_t launcher_hwnd,
                                   pipela::ui::dock::UiDockPhase phase) {
    if (phase == pipela::ui::dock::UiDockPhase::Client && game_hwnd &&
        pipela::core::win32::isWindow(game_hwnd)) {
        if (pipela::core::win32::isWindowMinimized(game_hwnd)) {
            return QString::fromUtf8("gmin:%1").arg(game_hwnd);
        }
        const auto [gw, gh] = clientSizeLogical(game_hwnd);
        const double ratio =
            static_cast<double>(std::max(1, gh)) / static_cast<double>(pipela::core::vision::kBaseHeight);
        const int dpi = static_cast<int>(std::lround(win32DpiScaleForHwnd(game_hwnd) * 96.0));
        return QString::fromUtf8("game:%1:%2:%3:%4:%5")
            .arg(game_hwnd)
            .arg(gw)
            .arg(gh)
            .arg(ratio, 0, 'f', 4)
            .arg(dpi);
    }
    if (launcher_hwnd && pipela::core::win32::isWindow(launcher_hwnd) &&
        !pipela::core::win32::isWindowMinimized(launcher_hwnd)) {
        const auto [lw, lh] = clientSizeLogical(launcher_hwnd);
        const int dpi = static_cast<int>(std::lround(win32DpiScaleForHwnd(launcher_hwnd) * 96.0));
        if (lw > 0 && lh > 0) {
            const double ratio =
                static_cast<double>(lh) / static_cast<double>(pipela::core::vision::kBaseHeight);
            return QString::fromUtf8("lsz:%1:%2:%3:%4:%5")
                .arg(launcher_hwnd)
                .arg(lw)
                .arg(lh)
                .arg(ratio, 0, 'f', 4)
                .arg(dpi);
        }
        return QString::fromUtf8("lonly:%1:%2").arg(launcher_hwnd).arg(dpi);
    }
    const int dpi = anchor_hwnd ? static_cast<int>(std::lround(win32DpiScaleForHwnd(anchor_hwnd) * 96.0))
                                : 96;
    return QString::fromUtf8("none:%1").arg(dpi);
}

void applyResolutionRichLabelFixed(QLabel* label, const QString& block_html, double design_scale) {
    if (label == nullptr) {
        return;
    }
    const QString inner = block_html.trimmed();
    if (inner.isEmpty()) {
        label->hide();
        return;
    }
    const QString cache_key =
        QString::fromUtf8("fixed:%1:%2").arg(inner).arg(design_scale, 0, 'f', 4);
    if (label->property("pipelaResFitCache").toString() == cache_key) {
        return;
    }
    label->setProperty("pipelaResFitCache", cache_key);
    label->setMaximumWidth(16777215);
    const double root_pt = QApplication::font().pointSizeF();
    const double scale = (root_pt / 11.0) * design_scale;
    const double chosen_pt = 10.5 * scale;
    QFont fnt = label->font();
    fnt.setPointSizeF(chosen_pt);
    label->setFont(fnt);
    label->setText(resLineWrap(inner, 0.0));
    label->setMinimumHeight(std::max(8, pipela::ui::theme::scalePxV(8, label->height() > 0 ? label->height() : 24)));
    label->show();
}

void applyResolutionRichLabelFit(QLabel* label, const QString& block_html, double avail_css_px,
                                 double design_scale) {
    if (label == nullptr) {
        return;
    }
    const QString inner = block_html.trimmed();
    if (inner.isEmpty()) {
        label->hide();
        return;
    }
    const double avail = std::max(40.0, avail_css_px);
    const double avail_fit = std::max(40.0, avail - static_cast<double>(scalePxV(4, 24)));
    const int avail_i = std::max(40, static_cast<int>(std::lround(avail)));
    const double root_pt = QApplication::font().pointSizeF();
    const QString cache_key = QString::fromUtf8("fit:%1:%2:%3:%4")
                                  .arg(inner)
                                  .arg(avail_i)
                                  .arg(design_scale, 0, 'f', 4)
                                  .arg(root_pt, 0, 'f', 3);
    if (label->property("pipelaResFitCache").toString() == cache_key) {
        return;
    }
    label->setMaximumWidth(avail_i);
    const double fit_limit = std::max(24.0, avail_fit * 0.99);
    const double scale = (root_pt / 11.0) * design_scale;
    const double base_pt = kResFitBaseDesignPt * scale;
    const double coarse_floor = kResFitMinDesignPt * scale;
    const double abs_floor = kResFitAbsoluteMinDesignPt * scale;
    double chosen_pt = abs_floor;
    double chosen_lsem = kResFitLetterSpacingEm[9];
    bool found = false;
    double pt = base_pt;
    {
        QFont f_quick = label->font();
        f_quick.setPointSizeF(pt);
        const QString w_quick = resLineWrap(inner, kResFitLetterSpacingEm[0]);
        if (measureResRichWidth(f_quick, w_quick) <= fit_limit) {
            chosen_pt = pt;
            chosen_lsem = kResFitLetterSpacingEm[0];
            found = true;
        }
    }
    while (!found && pt >= coarse_floor - 1e-9) {
        for (double lsem : kResFitLetterSpacingEm) {
            QFont fnt = label->font();
            fnt.setPointSizeF(pt);
            if (measureResRichWidth(fnt, resLineWrap(inner, lsem)) <= fit_limit) {
                chosen_pt = pt;
                chosen_lsem = lsem;
                found = true;
                break;
            }
        }
        if (!found) {
            pt -= kResFitPtStep;
        }
    }
    if (!found) {
        pt = std::min(base_pt, coarse_floor);
        while (!found && pt >= abs_floor - 1e-9) {
            for (double lsem : kResFitLetterSpacingEm) {
                QFont fnt = label->font();
                fnt.setPointSizeF(pt);
                if (measureResRichWidth(fnt, resLineWrap(inner, lsem)) <= fit_limit) {
                    chosen_pt = pt;
                    chosen_lsem = lsem;
                    found = true;
                    break;
                }
            }
            if (!found) {
                pt -= kResFitPtStepFine;
            }
        }
    }
    QFont fnt = label->font();
    fnt.setPointSizeF(chosen_pt);
    label->setFont(fnt);
    label->setText(resLineWrap(inner, chosen_lsem));
    label->setMinimumHeight(std::max(8, scalePxV(8, label->height() > 0 ? label->height() : 24)));
    label->setProperty("pipelaResFitCache", cache_key);
    label->show();
}

}  // namespace pipela::ui::theme
